### ISSUE TITLE
Create HDT files from RDF

### ISSUE THREAD

**Issue by donpellegrino** [KonradHoeffner/hdt#37]
It would be a valuable enhancement to create HDT files from RDF text serialization formats. A pure Rust implementation would be ideal, but a Rust wrapper over [hdt-cpp](https://github.com/rdfhdt/hdt-cpp) would be a convenient alternative.

**Comment by KonradHoeffner** [comment]
I agree that this would be very useful as hdt-cpp can be a bit difficult to install (except when using it over Docker but not everyone has that installed).
However I estimate that this would be a lot of work, as this library currently is read-only and has no HDT creation capability at all.
Unfortunately I cannot spend much time for HDT right now so successfully implementing this would probably require outside assistance, e.g. via a pull request.

**Comment by KonradHoeffner** [comment]
That sounds great! I have no experience with Rust interfacing C++ though and hope that it is not too difficult to handle the direct and transitive dependencies of hdt-cpp, do those then need to be wrapped as well?

**Comment by donpellegrino** [comment]
I have no idea yet. Based on "[Rust for Rustaceans](https://rust-for-rustaceans.com)," "Chapter 11: Foreign Function Interfaces," it looks like https://crates.io/crates/cxx is a good starting point. I still need to read through the cxx documentation.

**Comment by KonradHoeffner** [comment]
Related pull request https://github.com/KonradHoeffner/hdt/pull/56.

**Comment by KonradHoeffner** [comment]
@donpellegrino: By the way, the other way around (HDT to RDF) is also worked on by pchampin in the Sophia CLI: https://github.com/pchampin/sophia-cli/pull/4

Thanks to the work of @GregHanson, writing HDT should be integrated soon as well.

**Comment by KonradHoeffner** [comment]
Converting NT to HDT is now working (see branch rdf2hdt-integrated2) but I'm still refactoring a bit until it is polished enough for a release, probably there will be an alpha or beta first.

**Comment by KonradHoeffner** [comment]
@donpellegrino This is now in the main branch but the API may change until the next release, I'm open for suggestions in case something is not easy enough to use .
Feel free to test it using the Hdt::read_nt function.
Many thanks to @GregHanson.
